File: Java/MethodInvocationRemotingLoggingTests/src/net/alastairwyse/methodinvocationremotingloggingtests/TcpRemoteSenderLoggingTests.java
/*
 * Copyright 2014 Alastair Wyse (http://www.oraclepermissiongenerator.net/methodinvocationremoting/)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.alastairwyse.methodinvocationremotingloggingtests;

import org.junit.Before;
import org.junit.Test;
import java.nio.*;
import static org.mockito.Mockito.*;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.*;
import net.alastairwyse.methodinvocationremoting.*;
import net.alastairwyse.operatingsystemabstraction.*;
import net.alastairwyse.applicationlogging.*;

/**
 * Unit tests for the logging functionality in class methodinvocationremoting.TcpRemoteSender.
 * @author Alastair Wyse
 */
public class TcpRemoteSenderLoggingTests {

    private ISocketChannel mockSocketChannel;
    private IApplicationLogger mockApplicationLogger;
    private TcpRemoteSender testTcpRemoteSender;
    private String testIpAddress = "127.0.0.1";
    private int testPort = 55000;
    
    @Before
    public void setUp() throws Exception {
        mockSocketChannel = mock(ISocketChannel.class);
        mockApplicationLogger = mock(IApplicationLogger.class);
        testTcpRemoteSender = new TcpRemoteSender(testIpAddress, testPort, 3, 10, 25, 10, mockApplicationLogger, mockSocketChannel);
    }
    
    @Test
    public void ConnectLoggingTest() throws Exception {
        when(mockSocketChannel.isConnected())
            .thenReturn(false)
            .thenReturn(true);
        
        testTcpRemoteSender.Connect();
        
        verify(mockApplicationLogger).Log(testTcpRemoteSender, LogLevel.Information, "Connected to /" + testIpAddress.toString() + ":" + testPort + ".");
        verifyNoMoreInteractions(mockApplicationLogger);
    }
    
    @Test
    public void DisconnectLoggingTest() throws Exception {
        when(mockSocketChannel.isConnected()).thenReturn(true);
        
        testTcpRemoteSender.Disconnect();
        
        verify(mockApplicationLogger).Log(testTcpRemoteSender, LogLevel.Information, "Disconnected.");
        verifyNoMoreInteractions(mockApplicationLogger);
    }
    
    @Test
    public void SendLoggingTest() throws Exception {
        ByteBuffer acknowledgementBuffer = ByteBuffer.allocate(1);
        
        when(mockSocketChannel.isConnected()).thenReturn(true);
        // As the read method passes a reference to a ByteBuffer, and data is written to that ByteBuffer, to correctly mock this a custom Answer implementation is required.
        //   An instance of the AcknowledgementAnswer class is used.
        when(mockSocketChannel.read(acknowledgementBuffer)).thenAnswer(new AcknowledgementAnswer((byte)0x06, 1));
        
        testTcpRemoteSender.Send("<TestMessage>Test message content</TestMessage>");
        
        verify(mockApplicationLogger).Log(testTcpRemoteSender, LogLevel.Information, "Message sent and acknowledged.");
        verifyNoMoreInteractions(mockApplicationLogger);
    }
    
    /**
     * Mock answer for the SocketChannel.Read() method, to simulate the reading of the message acknowledgement byte
     * @author Alastair Wyse
     */
    private class AcknowledgementAnswer implements Answer<Integer> {

        private byte byteToWrite;
        private Integer returnValue;
        
        /**
         * Initialises a new instance of the AcknowledgementAnswer class.
         * @param byteToWrite  The single byte to return when the Read() method is called.
         * @param returnValue  The value to return.  Should be 1 for the success case.
         */
        public AcknowledgementAnswer(byte byteToWrite, Integer returnValue) {
            this.byteToWrite = byteToWrite;
            this.returnValue = returnValue;
        }
        
        @Override
        public Integer answer(InvocationOnMock invocation) throws Throwable {
            ByteBuffer returnByteBuffer = (ByteBuffer)invocation.getArguments()[0];
            returnByteBuffer.put(byteToWrite);

            return returnValue;
        }
    }
}

File: C#/MethodInvocationRemotingLoggingTests/TcpRemoteSenderLoggingTests.cs
/*
 * Copyright 2014 Alastair Wyse (http://www.oraclepermissiongenerator.net/methodinvocationremoting/)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Generic;
using System.Text;
using NUnit.Framework;
using NMock2;
using MethodInvocationRemoting;
using OperatingSystemAbstraction;
using ApplicationLogging;
using ApplicationMetrics;

namespace MethodInvocationRemotingLoggingTests
{
    //******************************************************************************
    //
    // Class: TcpRemoteSenderLoggingTests
    //
    //******************************************************************************
    /// <summary>
    /// Unit tests for the logging functionality in class MethodInvocationRemoting.TcpRemoteSender.
    /// </summary>
    [TestFixture]
    public class TcpRemoteSenderLoggingTests
    {
        private Mockery mocks;
        private ITcpClient mockTcpClient;
        private INetworkStream mockNetworkStream;
        private IApplicationLogger mockApplicationLogger;
        private TcpRemoteSender testTcpRemoteSender;
        private System.Net.IPAddress testIpAddress;
        private int testPort = 55000;

        [SetUp]
        protected void SetUp()
        {
            mocks = new Mockery();
            mockTcpClient = mocks.NewMock<ITcpClient>();
            mockNetworkStream = mocks.NewMock<INetworkStream>();
            mockApplicationLogger = mocks.NewMock<IApplicationLogger>();
            testIpAddress = System.Net.IPAddress.Parse("127.0.0.1");
            testTcpRemoteSender = new TcpRemoteSender(testIpAddress.ToString(), testPort, 3, 10, 25, 10, mockApplicationLogger, new NullMetricLogger(), mockTcpClient);
        }

        [Test]
        public void ConnectLoggingTest()
        {
            using (mocks.Ordered)
            {
                Expect.Once.On(mockTcpClient).GetProperty("Connected").Will(Return.Value(false));
                Expect.Once.On(mockTcpClient).Method("Connect");
                Expect.Once.On(mockApplicationLogger).Method("Log").With(testTcpRemoteSender, LogLevel.Information, "Connected to " + testIpAddress.ToString() + ":" + testPort + ".");
                Expect.Once.On(mockTcpClient).GetProperty("Connected").Will(Return.Value(true));
            }

            testTcpRemoteSender.Connect();

            mocks.VerifyAllExpectationsHaveBeenMet();
        }

        [Test]
        public void DisconnectLoggingTest()
        {
            using (mocks.Ordered)
            {
                Expect.Once.On(mockTcpClient).GetProperty("Connected").Will(Return.Value(true));
                Expect.AtLeastOnce.On(mockTcpClient);
                Expect.Once.On(mockApplicationLogger).Method("Log").With(testTcpRemoteSender, LogLevel.Information, "Disconnected.");
            }

            testTcpRemoteSender.Disconnect();

            mocks.VerifyAllExpectationsHaveBeenMet();
        }

        [Test]
        public void SendLoggingTest()
        {
            using (mocks.Ordered)
            {
                Expect.Once.On(mockTcpClient).GetProperty("Connected").Will(Return.Value(true));
                Expect.Once.On(mockTcpClient).Method("GetStream").WithNoArguments().Will(Return.Value(mockNetworkStream));
                Expect.Once.On(mockNetworkStream).Method("Write");
                Expect.Once.On(mockTcpClient).GetProperty("Available").Will(Return.Value(1));
                Expect.Once.On(mockNetworkStream).Method("ReadByte").WithNoArguments().Will(Return.Value(6));
                Expect.Once.On(mockApplicationLogger).Method("Log").With(testTcpRemoteSender, LogLevel.Information, "Message sent and acknowledged.");
            }

            testTcpRemoteSender.Send("<TestMessage>Test message content</TestMessage>");

            mocks.VerifyAllExpectationsHaveBeenMet();
        }
    }
}
